### ISSUE TITLE
Screensplit of Psycho World shows garbage

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1310]
On real MSX it's perfectly clean... See also #470 where this is already mentioned. But I think it's better to make separate tickets. The Hydefos issue (from the same developer) might be the same thing.

**Comment by grauw** [comment]
Also on my turboR GT the game looked clean.

In openMSX one line past the game view is flickering on and off. This suggests that the display disable is set right on the edge of where it’s sampled. It flickers due to variations in interrupt response timing depending on the duration of the instruction that happens to be executing when the IRQ occurs. Indeed, I confirmed that by giving just 10 cycles of extra time the split becomes clean.

Some possible explanations:

1. The display enable flag (BL) is sampled later on a real V9938
2. The horizontal line interrupt occurs too late
3. The CPU interrupt response takes longer than on a real machine

I made a test program to measure where the BL flag takes effect. It alternates between toggling the background colour and blanking every 228 cycles, and can offset the cycles by 1 wit the cursor keys. I measured that it is 44-45 Z80 CPU cycles to the left of the active display, which about matches up with the start of the synchronize signal period. So this matches the openMSX implementation.

**Comment by grauw** [comment]
Let me also write down the general outline of the code involved with the interrupts;

Psycho World has 2 line interrupts, one on line 142 and one on line 28. The interrupt handler code is at 7376H (VBlank interrupt), 740CH (line 142 interrupt) and 748FH (line interrupt).

On line 142 (address 740CH), the split in question, it does this:

```
Line142_Handler:
    r#1 = 22H             ; display disable
    r#15 = 01H            ; s#1
    read s#1              ; clear line interrupt
    r#15 = 00H            ; s#0
    r#0 = 06H             ; disable h.int, screen 5
    r#2 = 7FH             ; pattern name table
    r#23 = 76H            ; display offset
    r#23 = 76H            ; display offset
    r#18 = 00H            ; display adjust
    r#9 = 80H             ; 212 lines
    r#5 = BFH             ; SAT
    r#6 = 85H             ; SPT
    set palette
    r#1 = (A097H)  ; display enable
    write to sprite attribute table (20 sprites)
    (0A081) = 02H
    (H.KEYI + 1) = VBlank_Handler
    pop & exit
```

On vertical blank (address 7376H) it does:

```
VBlank_Handler:
    r#19 = (A083H) + 28   ; next interrupt line
    r#0 = 14H             ; screen 4, enable h.int
    r#18 = (A082H)        ; display adjust
    r#2 = (A084H)         ; pattern name table
    r#23 = (A083H)        ; display offset 
    (H.KEYI + 1) = Line28_Handler
    r#5 = 3FH             ; SAT
    r#6 = 07H             ; SPT
    (A08AH) = (A083H)
    set palette / fade (?)
    read s#0              ; clear vblank interrupt
    ei
    misc game / music stuff (?)
    di
    pop & exit
```

On line 28 (address 748FH) it does:

```
Line28_Handler:
    r#15 = 01H            ; s#1
    read s#1              ; clear line interrupt
    r#15 = 00H            ; s#0
    r#19 = (A083H) + 142  ; next interrupt line
    r#0 = 14H             ; screen 4, enable h.int
    write to sprite attribute table
    (H.KEYI + 1) = Line142_Handler
    pop & exit
```

**Comment by grauw** [comment]
### Test programs

Attached two [test programs](https://github.com/openMSX/openMSX/files/5748689/bl-tests.zip). Both of these run in SCREEN 1.

#### VDPBLANK

Measures where the BL flag takes effect. It alternates between toggling the background colour and blanking every 228 cycles, and can offset the cycles by 1 with the cursor keys. It shows a matching ASCII character in the top left which can be cross referenced with a [character set table](https://www.msx.org/wiki/images/6/66/MSX_font-Int.png) to determine the exact amount.

This is the one I mentioned in my first comment. 

#### HINTTOBL

Measures the time between the line interrupt and BL being sampled by the VDP and taking effect. It will stop showing a blank line when the wait becomes too long. The number of cycles until (excluding) the 11th cycle of the OUT (C) instruction is shown on screen for quick referencing. Due to HALT alignment (5 cycles), the last last 4 lines flicker. To determine the exact time, take the maximum value.

### Test results

Yamaha CX5MII (real)

- 242 CPU cycles between line interrupt and BL flag.
- BL is sampled 19 CPU cycles to the left of active display

Yamaha CX5MII (openMSX)

- 217 CPU cycles between line interrupt and BL flag.
- BL is sampled 43 CPU cycles to the left of active display

So in conclusion, BL is sampled ~24 CPU cycles too early in openMSX, or ~144 VDP cycles. This is right in the middle of the gap in the left erase signal as seen in the [timing chart](http://map.grauw.nl/articles/vdp-vram-timing/vdp-timing-v2.png), which makes a lot of sense.

Probably not only BL, but also the mode and SPD flags activate at this point in time, I have not confirmed this but it seems to fit more logically in the VDP timing chart.

The reason why I did not already find this out earlier is because I confused the MSX character code of Ü with ü.

**Comment by grauw** [comment]
I also tested the M1 bit, with a simple change to HINTTOBL (see below), the results are the same.

```
EnableDisable:
	ld b,1 | 80H
	ld a,(RG1SAV)
	set 4,a
	out (c),a
	out (c),b  ; a 38 cycle window
	res 4,a
	out (c),a
	out (c),b
	ret
```

**Comment by grauw** [comment]
I also tested the horizontal adjust register (r#18), with a simple change to HINTTOBL (see below), the results are the same.

```
EnableDisable:
	ld b,18 | 80H
	ld a,8
	ld a,8
	ld a,8
	out (c),a
	out (c),b  ; a 38 cycle window
	res 3,a
	out (c),a
	out (c),b
	ret
```

p.s. I notice that on my real MSX actually the display below the threshold point starts shaking vertically, as if some lines are skipped. Could be either my IPS display or OSSC not liking timing changes caused by adjust, or the VDP line counter skipping if the timing is exactly right. (I am dreaming of horizontal overscan now.)

**Comment by grauw** [comment]
I notice in openMSX when SET ADJUST (r#18) is not 0, it affects the results. On the real MSX it does not.

If r#18 is 7, openMSX reads 246 CPU cycles. If r#18 is 8, it reads 236 CPU cycles. The real MSX stays on 242.

**Comment by grauw** [comment]
I also tested the SPD bit, with a simple change to HINTTOBL (see below), the results are the same.

```
EnableDisable:
	ld b,8 | 80H
	ld a,(RG8SAV)
	set 1,a
	out (c),a
	out (c),b  ; a 38 cycle window
	res 1,a
	out (c),a
	out (c),b
	ret
```

I updated [the test program](https://github.com/openMSX/openMSX/files/5748689/bl-tests.zip) to draw sprites as well so that the effect of the above change can be seen.

I also note that in openMSX, sprites are not actually enabling per-line (see #1104). Additionally, they should not re-enable instantly, there is one extra line of delay because it needs to re-read the sprite data.

**Comment by grauw** [comment]
Patch:

```diff
diff --git a/src/video/VDP.cc b/src/video/VDP.cc
index f578b1685..eae47c808 100644
--- a/src/video/VDP.cc
+++ b/src/video/VDP.cc
@@ -1223,8 +1223,11 @@ void VDP::changeRegister(byte reg, byte val, EmuTime::param time)
 
 void VDP::syncAtNextLine(SyncBase& type, EmuTime::param time)
 {
-	int line = getTicksThisFrame(time) / TICKS_PER_LINE;
-	int ticks = (line + 1) * TICKS_PER_LINE;
+	// The processing of a new line starts in the middle of the left erase,
+	// ~144 cycles after the sync signal. Adjust affects it. See issue #1310.
+	int offset = 144 + (horizontalAdjust - 7) * 4;
+	int line = (getTicksThisFrame(time) + TICKS_PER_LINE - offset) / TICKS_PER_LINE;
+	int ticks = line * TICKS_PER_LINE + offset;
 	EmuTime nextTime = frameStartTime + ticks;
 	type.setSyncPoint(nextTime);
 }
```

**Comment by grauw** [comment]
I checked and this also fixes Hydefos and Space Manbow.

**Comment by MBilderbeek** [comment]
It also fixes the Woei demo and the bottom part of the Dawn of Time demo (copper part).

**Comment by m9710797** [comment]
I reviewed/pushed grauw's patch. See revision 547064a2e6578b5.
@grauw: Thanks a lot!

**Comment by grauw** [comment]
Illustrational picture:

![afbeelding](https://user-images.githubusercontent.com/772052/121530851-2257fd00-c9fe-11eb-9df5-c9a846556c65.png)

Right around there is where BL, SPD and M1-M5 are latched.

**Comment by artrag** [comment]
Awesome finding! Great work Grauw!
